walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -33,13 +35,48 @@
     Collection<TimeRange> optResults = queryInternal(events, requestWithOptionals);
     
     if ((optResults.isEmpty()) && (!request.getAttendees().isEmpty())) {
-      return queryInternal(events, request);
+      return queryBFS(events, requestWithOptionals, request.getAttendees(),
+          request.getOptionalAttendees());
     }
     else {
       return optResults;
     }
   }
 
+  /* Follow a BFS, removing optional attendees gradually and returning when you find a timerange
+  which accomodates the remaining optional attendees and the required attendees. If you never
+  find such a timerange, return the results just accounting for required attendees. */
+  public Collection<TimeRange> queryBFS(Collection<Event> events, MeetingRequest request, Collection<String> reqAttendees,
+        Collection<String> optAttendees) {
+    Queue<Collection<String>> queue = new ArrayDeque<Collection<String>>();
+    queue.add(optAttendees);
+
+    while (!queue.isEmpty()) {
+      Collection<String> curOptAttendees = queue.remove();        
+      Collection<String> curAttendees = new ArrayList<String>();
+      curAttendees.addAll(curOptAttendees);
+      curAttendees.addAll(reqAttendees);
+
+      MeetingRequest curRequest = new MeetingRequest(curAttendees, 
+          request.getDuration());
+      Collection<TimeRange> results = queryInternal(events, curRequest);
+      if (!results.isEmpty()) {
+        return results;
+      }
+      
+      else {
+        for (String optAttendee : curOptAttendees) {
+          Collection<String> newOptAttendees = new ArrayList<String>();
+          newOptAttendees.addAll(curOptAttendees);
+          newOptAttendees.remove(optAttendee);
+          queue.add(newOptAttendees);
+        }
+      }
+    }
+    MeetingRequest reqRequest = new MeetingRequest(reqAttendees, request.getDuration());
+    return queryInternal(events, reqRequest);
```

Comment:
Just wondering, are these two lines required? Or would we eventually hit the "empty optional attendee" state with lines 69-72 anyway?

Reply:
I definitely agree that these lines are unnecessary. I only included them to avoid a "missing return statement" error. Is there a better way to avoid this error? Should I return an empty arrayList instead?

Reply:
Oh I see. How about declaring `results` outside the scope of the while loop, and just having `return results` down here, that should work?

Reply:
Sound perfect! Thank you!